### PR TITLE
switch languages without an app restart

### PR DESCRIPTION
**Pull request by Docteh**
Hello, Looks like anything except the GUI.log can be re-translated on the fly.

Give it a go. I'll fix the commit with a revision to `locales/en/messages.json` if needed. But maybe we can just drop the string that says that a restart is required.

`<span i18n="userLanguageSelect" class="i18n-replaced">Sprache (Das Programm muss zur Änderung neu gestartet werden)</span>`

### CODE REVIEW

**Comment by mikeller** [comment]
Except that the builds are failing. :-(

**Comment by Docteh** [comment]
Whoops, not the only bug.
image
 
Also fixed.

**Comment by McGiverGim** [comment]
I did a similar feature when I made the translation system, and @mikeller asked to remove this live language change because not all is changed. Some combos that are loaded in the js and other thinks are not translated. I suggested some workarounds but it was some more little work for the developers.

I will look at the discussion to see if I can find it...

**Comment by McGiverGim** [comment]
Here is the original PR.

**Comment by Docteh** [comment]
Thanks I'll take a closer look at what doesn't update. If we can get the last percentage fixed I think we can make a strong case going forward for people to keep things translatable on the fly. I figure that the log could be left alone just because of how logs work. Sometimes these things need to sneak up in stages. 

I should benchmark the retranslation compared to asking it to translate new items in the document.

**Comment by mikeller** [comment]
@McGiverGim: Good point, I remember now. My point still stands that, if this requires extra coding effort and understanding from people wanting to contribute to the configurator, then this disadvantage outweighs the benefits we get from it. After all, for the majority of users, selecting the language will be a one-off operation, and they won't have any need to switch the language again after selecting the one they want, so having to restart the application for this is only a very minor inconvenience.

@Docteh: Agreed that leaving the log in whatever language it was generated is acceptable.

**Comment by Docteh** [comment]
Here is an example of an item that is tricky to retranslate. The two drop downs above it are easier.

Does it really need to say "Choose version for ....."? I figure just selecting the latest version should be fine. Or just make it unspecific.
image

**Comment by mikeller** [comment]
@Docteh:

> Does it really need to say "Choose version for ....."? I figure just selecting the latest version should be fine. Or just make it unspecific.

That again for me turns into a matter of deciding what is more important - the capability to change languages without a restart, or the capability to add translations to whatever we want without restriction, and without having to think about what is working and what isn't.

**Comment by Docteh** [comment]
How about a compromise?

We could only change the language if the GUI doesn't think its connected.

Leave the message stating that a restart is required.

Right now if something on a tab doesn't switch languages, you can just go to a different tab, and then come back.

**Comment by mikeller** [comment]
Interesting point, I never looked at it this way, but since tabs are reloaded entirely on every change this will work.

Or take this in a different direction, and put the language drop down into the landing tab (this makes it easier to discover for somebody not fluent in the display language as well).

**Comment by Docteh** [comment]
Put the language switcher on the landing page? I like that idea. That'll be my next move.

**Comment by Docteh** [comment]
Looking for input, this probably needs a touch of design work.  I thought about throwing the selector widget into the page, and ended up researching language choosers for a bit. I like having the list.

image

**Comment by mikeller** [comment]
I like it - it shows the option in your language / alphabet right on startup - doesn't go much more straightforward than that.
Speaking from experience of clicking through a plethora of drop downs filled with Chinese characters in search of the one listing 'English', I think this is great.
I'd just add a blurb up top explaining that this is to select the language, and will be stored permanently. And maybe change 'Default' to 'System Default', or add a blurb explaining that this is whatever language your system is in (if available).

File: locales/en/messages.json [hunk 1]
```diff
@@ -72,8 +72,12 @@
     "userLanguageSelect": {
         "message": "Language (need to restart the application for the changes to take effect)"
     },
```
Comment (McGiverGim) [inline locales/en/messages.json]: If we are removing the combo from options this needs to be removed too.

File: src/js/localization.js [hunk 2]
```diff
@@ -36,6 +36,28 @@ i18n.init = function(cb) {
                     }
             });
     });
+    // This function should do the same things that the i18n.localizePage function below does.
+    i18next.on('languageChanged', function (newLang) {
```
Comment (McGiverGim) [inline src/js/localization.js]: I don't think this is the best way to do it. The code is very similar to function `localizePage`, I think is better to use the same function when is the first time that we localize the page or when we are doing it after each change of language. Maybe pass a parameter to a third method that can be used for both functions. In this way we will have only one place to maintain if needed.

File: src/js/tabs/landing.js [hunk 3]
```diff
@@ -2,19 +2,60 @@
...
+        var languageSelected = element.attr('lang');
+        if (newLang == languageSelected) {
+          element.removeClass('selected_language');
+          element.addClass('selected_language');
+        } else {
+          element.removeClass('selected_language');
+        }
+      });
+    }
+    ConfigStorage.get('userLanguageSelect', function (result) {
```
Comment (McGiverGim) [inline src/js/tabs/landing.js]: This was done like this in the "original" version, but now that we are modifying it, maybe is better to expose the method of `localization.js` that returns the language selected by the user. In this way the string `userLanguageSelect` will not need to be used outside the `localization.js` file. I don't know if the current implementation of this method can be used as is, but we can refactor it.

**Comment by Docteh** [comment]
image

For the languages, not sure if I want like `English (en)` or just `English`. Leaning towards just `English`. I'll remove eg `languagebr_en` from the english locale file if we go that way otherwise its useful.

**Comment by McGiverGim** [comment]
The locales were there to give "some order" to the texts. They are ordered by locale. I don't know if now, without them, it seems more only a mesh of languages...

**Comment by mikeller** [comment]
I am a bit on the fence about adding the locale for every language to the list - my sense of ordering likes it, my sense of curiosity likes to be able to see what the languages not using the latin alphabet are, but my sense of aestethics just loves the simplicity of the list of languages without any distractions.

But one thing I think we need to address is that currently the cursor when hovering over the language names is text selection - this makes it harder than needed to discover that they are clickable.

**Comment by Docteh** [comment]
I switched the spans out for links, so they'll be more obviously clickable. 

We could also do something to the text on hover. Like underline, or a color change.

File: locales/fake/messages.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+    "language_default_pretty": {
```
[the remark this answers is not in the document]
Comment (mikeller) [reply locales/fake/messages.json]: Or to be more precise: What is this used for?

[the remark this answers is not in the document]
Comment (Docteh) [reply locales/fake/messages.json]: In English language_default_pretty is fully defined: "System Default ($t(detectedLanguage))"

This works as a fallback as many languages have language_default translated.

messing with fallback might be a way to rename a key, and not lose the translated content. 

tabLanding
defaultChangelogAction
defaultPrivacyPolicyAction
tabHelp
tabFirmwareFlasher

**Comment by Docteh** [comment]
Looks like I did something slightly wrong. I'll try to fix. 

Let me know on the fallback fakery.

File: locales/fake/messages.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikeller) [reply locales/fake/messages.json]: Not sure I understand this right, but the way `fallbackLng` is used it looks now like `fake` is the only language that will ever be used as fallback language. This is probably not what we want.

Also, should this mythical language maybe be called `fallback` or similar instead of `fake`?

[the remark this answers is not in the document]
Comment (Docteh) [reply locales/fake/messages.json]: I believe the fallback is to english. We can specify multiple fallback languages. So for `defaultDonateBottom` on the landing page gets english when using german. (my change to the link ended up wiping out the translations).

[the remark this answers is not in the document]
Comment (mikeller) [reply locales/fake/messages.json]: Have a look at the code. Only the first element of `fallbackLng` seems to ever be used. Is it possible that English is a hardcoded default fallback for the framework we are using?

[the remark this answers is not in the document]
Comment (Docteh) [reply locales/fake/messages.json]: In i18next or in our code? We **are** injecting information (resources) into the first fallbackLng, but I figured that would be okay to ignore. I figure it'd be an annoying be problem to throw out a language while running, even if its a fake one.

For this screenshot I made did `fallbackLng: ['fallback','de','en'],` and then loaded polish translation manually `i18next.changeLanguage('pl')`
image

[the remark this answers is not in the document]
Comment (mikeller) [reply locales/fake/messages.json]: Sorry, my bad, I'd assumed `fallbackLng` was an option that we'd added (dodgy name and all).

File: src/js/tabs/landing.js [hunk 5]
```diff
@@ -2,18 +2,51 @@
 
 TABS.landing = {};
 TABS.landing.initialize = function (callback) {
-    var self = this;
+  var self = this;
```
Comment (mikeller) [inline src/js/tabs/landing.js]: This file will need to have the indentation adjusted at some point (not in this pull request though).

**Comment by mikeller** [comment]
@McGiverGim: Is this ok now?

**Comment by McGiverGim** [comment]
My review yes, now has been fixed. But I don't understand all this fallback language. @Docteh can you explain why it is needed? I see that you have translations that recommend to translate in other place, translations that are in the fallback...

File: locales/en/messages.json [hunk 6]
```diff
@@ -574,13 +586,15 @@
         "message": "We also have a <a href=\"https://www.facebook.com/groups/betaflightgroup/\" target=\"_blank\">Facebook Group</a>.<br />Join us to get a place to talk about Betaflight, ask configuration questions, or just hang out with fellow pilots."
     },
     "defaultChangelogAction": {
-        "message": "Changelog"
+        "message": "Changelog",
+        "description": "Please translate tabChangelog instead"
```
Comment (McGiverGim) [inline locales/en/messages.json]: Why is this needed? Do you want that the translators let this as is and go to the other translation and modify the content (that is a reference) with the real translation? This will not happen for sure...

**Comment by McGiverGim** [comment]
It seems it has been done in this way to maintain some actual translations. If they are not valid/needed the correct way is to remove/change them. I don't have too much time now to look deeply to the PR, so if it is needed for certain reasons sorry, I have not see them.

File: locales/en/messages.json (hunk 6 above)
Comment (Docteh) [reply locales/en/messages.json]: I may be making assumptions based on a misunderstanding.

This PR adds a string tabChangelog, which will show as a new string requiring translation

I can't predict if we're too close to release to expect that it will be translated, but since its the same content as defaultChangelogAction, we can use that.

At some point we remove defaultChangelogAction from `fallback` and `en`, and I think the removal from `en` will remove it from the other languages automatically.

Comment (McGiverGim) [reply locales/en/messages.json]: The translations works in this way:
- If we change/add a translation, the translators will be notified to do the job. If some of them don't do it, it must fallback to English.
- We let two or three days before releasing to let the translators finish the language. If some of them don't do it, it will be English.
- Adding the fallback messages file, what we have is what I asked in the other PR: a new file where we can put the constants messages (it will not be translated). I don't know if this is what we want.

Don't worry if we lose a translation (they are two little strings). We must do code easy to maintain and if we make this hacks, they will last forever.

Comment (Docteh) [reply locales/en/messages.json]: calling it constants kind of bothered me. something in the direction of `locales/not_translated/messages.json` would be a bit more clear. 

For these hacks we could commit to a time to rip them out.

Comment (mikeller) [reply locales/en/messages.json]: I agree with @McGiverGim's point - we are not in a situation where we can expect all of our translations to be 100% complete - after all they are done by volunteers, and we can't delay releases indefinitely in order to give them time to complete translations.
In this environment, it is probably better to not introduce more complexity in order to enable workarounds for translation keys that are changed - simply introduce the new key, and wait for the translators to do their work in the time they need, just like we are already doing for new translation keys that are added.

**Comment by Docteh** [comment]
Oh, probably should have commented that the `fallback` modifications are pulled.
